### PR TITLE
Initial docs

### PR DESCRIPTION
**Pull request by uhbif19**

### CODE REVIEW

File: docs/goals_and_soa.md [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+5. Generally production ready (production-ready)
+
+## Not in scope at all
+
+* Composable inter-DApp interactions and DApp migrations
+* Scheduled actions
+
+## Not in scope for current Catalyst funding
+
+* CTL and overall frontend transaction construction model
```
Comment (euonymos) [inline docs/goals_and_soa.md:32]: How does it differ from mentioned earlier "offchain Tx construction and indexing backend"? Do we cover any other means to generate off-chain code?

File: docs/goals_and_soa.md [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+There is a list of known security vulnerabilities,
+for which naive Plutus implementation is very often prone to.
+Our implementation makes them impossible to happen,
+taking that burden from developers and auditors.
+
+* Double satisfaction
+    * Shown twice in CNS Audit
+    * Shown once in Frada audit (3.2.7)
+* Insufficient staking control
+    * Shown once in CNS Audit - https://github.com/mlabs-haskell/cns-plutus-audit/issues/24
```
Comment (euonymos) [inline docs/goals_and_soa.md:74]: The link is not accessible for everyone (including me).

File: docs/goals_and_soa.md [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+
+@todo #3: coin-selection, tests, retrying and errors
+
+### Comon backend features
+
+There is a list of common tasks shared by multiple backends,
+which could be tackled generically in our framework.
+
+* Parsing transaction back to state-machine transition
+is required for delegated architecures,
```
Comment (euonymos) [inline docs/goals_and_soa.md:215]: What are delegated architectures? Have them got to do with Hydra delegation topology.

File: docs/goals_and_soa.md [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+
+### General Haskell verification tools
+
+We plan to use first two tools in our implementation.
+They both are tested in real prjects and documented well.
+
+Other tools are not applicable to our project.
+
+1. quickehck-dynamic
+2. SMV
```
Comment (euonymos) [inline docs/goals_and_soa.md:359]: What's SMV?

File: docs/goals_and_soa.md (hunk 3 above)
Comment (uhbif19) [reply docs/goals_and_soa.md:215]: Delegated architecture is one where each client only contacts chain, in that case L2, thought Delegate servers. That is simplest and only architecture working right now for Hydra. We spawn bunch of Hydra Nodes, each containing Delegate server accepting client requests. I am saying about old Hydra Auction or probably any Hydra baked project for now.

This Delegate server should keep track of Hydra state overall and Tx pushed on it specifically. Which gives a need to parse Tx from intended DApp protocol, to keep its current state. For example: Auction has CurrentBid state, which delegate server should be aware of. This means some DApp Tx-parsing should be present. We aim to give generic utils for this.

File: docs/goals_and_soa.md (hunk 4 above)
Comment (uhbif19) [reply docs/goals_and_soa.md:359]: This is very usable Haskell SMT interface. It provides both DSL for SMT constraints encoding and solving and soe more complicated static analysis DSL examples on top of that.

File: docs/goals_and_soa.md [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+
+Another important restriction,
+is that PAB state machines always produce state-thread-token contract.
+This makes impossible to encode tree like UTxO based data-structures,
+which are used for encoding all kinds of public registry logic.
+Also it prohibits any kind of non-trivial state-thread-token logic,
+like optimization by sharing (used for example in `hydra-auction`)
+and multiple commiters schemes (used in `hydra`).
+
+@todo #3: Write more about PAB issues with emulator, packaging and Plutus
```
[the remark this answers is not in the document]
Comment (uhbif19) [reply docs/goals_and_soa.md:274]: Please tell me about main problems known. You want give links or write text yourself?

File: docs/goals_and_soa.md (hunk 2 above)
Comment (uhbif19) [reply docs/goals_and_soa.md:74]: Yes. I would ask about it later, but I think they are not published code/audit yet.

File: docs/goals_and_soa.md (hunk 1 above)
Comment (uhbif19) [reply docs/goals_and_soa.md:32]: > indexing backend

It actually get out of scope for Milestones accepted.

> Do we cover any other means to generate off-chain code?

There would be no off-chain code generation. Only typeclasses.

**Pull request merged**